src/lib.rs

```diff
@@ -0,0 +1,534 @@
+}
+
+// TODO: A better name for this struct
+/// A struct for storing states of TimeoutChecker specific to a test instance
+#[derive(Clone, Debug)]
+struct TimeoutCheckerTestInstance {
+    test_start_time: Instant,
+    expected_iter: u64,
+
+    completed_iter: u64,
+    checkpoint: u64,
+    num_checks_completed: u128,
+    checking_interval: Duration,
+}
+
+impl Memtester {
+    // TODO: Memtester without given base_ptr, ie. take care of memory allocation as well
+
+    /// Create a Memtester containing all test types in random order
+    pub fn all_tests_random_order(args: MemtesterArgs) -> Memtester {
+        let mut test_types = vec![
+            MemtestType::TestOwnAddress,
+            MemtestType::TestRandomVal,
+            MemtestType::TestXor,
+            MemtestType::TestSub,
+            MemtestType::TestMul,
+            MemtestType::TestDiv,
+            MemtestType::TestOr,
+            MemtestType::TestAnd,
+            MemtestType::TestSeqInc,
+            MemtestType::TestSolidBits,
+            MemtestType::TestCheckerboard,
+            MemtestType::TestBlockSeq,
+        ];
+        test_types.shuffle(&mut thread_rng());
+
+        Self::from_test_types(args, test_types)
+    }
+
+    /// Create a Memtester with specified test types
+    pub fn from_test_types(args: MemtesterArgs, test_types: Vec<MemtestType>) -> Memtester {
+        Memtester {
+            test_types,
+            timeout: args.timeout,
+            require_memlock: args.require_memlock,
+            allow_working_set_resize: args.allow_working_set_resize,
+            allow_mem_resize: args.allow_mem_resize,
+            allow_multithread: args.allow_multithread,
+            allow_early_termination: args.allow_early_termination,
+        }
+    }
+
+    /// Run the tests, possibly after locking the memory
+    pub fn run(&self, memory: &mut [usize]) -> anyhow::Result<MemtestReportList> {
+        // TODO: Should have a minimum memory length so that we don't UB when `memory.len()` is too small
+        let mut timeout_checker = TimeoutChecker::new(Instant::now(), self.timeout);
+
+        // TODO: the linux memtester aligns base_ptr before mlock to avoid locking an extra page
+        //       By default mlock rounds base_ptr down to nearest page boundary
+        //       Not sure which is desirable
+
+        if self.require_memlock {
+            #[cfg(windows)]
+            let working_set_sizes = if self.allow_working_set_resize {
+                Some(
+                    windows::replace_set_size(size_of_val(memory))
+                        .context("Failed to replace process working set size")?,
+                )
+            } else {
+                None
+            };
+
+            let (memory, mlocked) = match memory_resize_and_lock(memory, self.allow_mem_resize) {
+                Ok(resized_memory) => (resized_memory, true),
+                Err(e) => {
+                    // TODO: Returning without restoring set size?
```

Comment:
Indeed -- I'm glad you're thinking about this kind of stuff! The best way to prevent this is for you to implement the `Drop` trait on some kind of "guard object". When you do that, `drop()` will be called automatically whenever the guard goes out of scope -- No need to manually worry about calling `restore_set_size` anymore. Below is code that will do what you need.

Change your code on Line 127 to this:

```Rust
let _working_set_resize_guard = if self.allow_working_set_resize {
    Some(
        windows::replace_set_size(size_of_val(memory))
            .context("failed to replace process working set size")?,
    )
} else {
    None
};
```

And add this to your `windows` module:

```Rust
#[derive(Debug)]
pub(super) struct WorkingSetResizeGuard {
    min_set_size: usize,
    max_set_size: usize,
}

pub(super) fn replace_set_size(memsize: usize) -> anyhow::Result<WorkingSetResizeGuard> {
    let (mut min_set_size, mut max_set_size) = (0, 0);
    unsafe {
        GetProcessWorkingSetSize(GetCurrentProcess(), &mut min_set_size, &mut max_set_size)
            .context("failed to get process working set")?;
        // TODO: Not sure what the best choice of min and max should be
        SetProcessWorkingSetSize(
            GetCurrentProcess(),
            memsize.saturating_mul(2),
            memsize.saturating_mul(4),
        )
        .context("failed to set process working set")?;
    }
    Ok(WorkingSetResizeGuard {
        min_set_size,
        max_set_size,
    })
}

impl Drop for WorkingSetResizeGuard {
    fn drop(&mut self) {
        unsafe {
            if let Err(e) = SetProcessWorkingSetSize(
                GetCurrentProcess(),
                self.min_set_size,
                self.max_set_size,
            ) {
                warn!("failed to restore process working set: {e}");
            }
        }
    }
}
```